### PR TITLE
Adding CleanShotX Extension

### PR DESCRIPTION
**Pull request by Aayush9029**
## Description

Extension enables users to speedup their CleanShotX workflow. Allowing screen capture, via raycast.

## Type of change

- New extension

## Screencast

https://user-images.githubusercontent.com/43297314/147637266-c8cd382c-50fb-417b-a322-cb04ddf38601.mov

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)

### CODE REVIEW

File: extensions/cleanshotx-extension/package.json
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/package.json:13]:
```suggestion
      "title": "Take screenshots",
      "subtitle": "CleanShot X",
```

File: extensions/cleanshotx-extension/package.json
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/package.json:4]:
```suggestion
  "name": "cleanshotx",
  "title": "Cleanshot X",
```

File: extensions/cleanshotx-extension/package.json
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/package.json:5]:
```suggestion
  "description": "Capture and record your screen!",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:91]:
```suggestion
              <OpenInBrowserAction title="Run Action" url={item.urlScheme} />
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:9]:
```suggestion
      subtitle: "Fullscreen screenshot",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:15]:
```suggestion
      subtitle: "Selection tool capture",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:21]:
```suggestion
      subtitle: "Previously selected area",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:27]:
```suggestion
      subtitle: "Window capture",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:33]:
```suggestion
      subtitle: "Timer based screen capture",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:39]:
```suggestion
      subtitle: "Scrolling capture mode",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:45]:
```suggestion
      subtitle: "Record entire screen",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:51]:
```suggestion
      subtitle: "Text recognition using OCR",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:57]:
```suggestion
      subtitle: "Annotate mode",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:63]:
```suggestion
      subtitle: "Open image from clipboard",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:69]:
```suggestion
      subtitle: "Toggle desktop icons visibility",
```

File: extensions/cleanshotx-extension/src/index.tsx
Comment (pernielsentikaer) [inline extensions/cleanshotx-extension/src/index.tsx:76]:
```suggestion
      subtitle: "Restore recently closed overlay",
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @Aayush9029 👋 

Awesome extension, I'm gonna use it myself for sure 💪 

I left a few comments for you to work:

- The extensions folder should be renamed to cleanshotx

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions 💯

**Comment by unnamedd** [comment]
Hello @Aayush9029,

Thanks for your contribution to our community, however I'm curious about what's the benefit your extension will have over the current implementation of the same in Script Commands. 

Here are some of the concerns I have:

- The extension is not bringing a different experience to the user in terms of feature.

- As far as I could see reading the code, you are not creating different commands for each CleanShot X action, you are implementing everything in the same list, which is bad compared to the Script Commands, because there, you can customize using an alias or defining a shortcut for a quick access.

You could argue that Extensions are easier to install than Script Commands, however, the great benefit to have different commands for each action isn't being used, which forces the user to access a list and then the action preferred.

I use myself many times a day the "Open From Clipboard" action and I have a shortcut defined for it. How your extension would make my life easier to access that action?

**Comment by pernielsentikaer** [comment]
> Thanks for your contribution to our community, however I'm curious about what's the benefit your extension will have over the current implementation of the same in Script Commands.

Again, why not have it in the store? It's much easier for the users to install from Store.

I think it makes perfectly sense to have it, assign a hot key and easy type what you want to achieve.

**Comment by unnamedd** [comment]
> > Thanks for your contribution to our community, however I'm curious about what's the benefit your extension will have over the current implementation of the same in Script Commands.
> 
> Again, why not have it in the store? It's much easier for the users to install from Store.
> 
> I think it makes perfectly sense to have it, assign a hot key and easy type what you want to achieve.
> 
> Sorry but I disagree with you

Sorry to disagree with you to Per, but my point isn't exactly in having it or not, but in the User Experience like I mentioned in my original message.

**Comment by pernielsentikaer** [comment]
> > Sorry but I disagree with you
> 
> Sorry to disagree with you too Per, but my point isn't exactly in having it or not, but in the User Experience like I mentioned in my original message.

Sorry if I misunderstood you, it's fine adding it to extensions atleast 😊 

I like the way it's implemented but both ways could work for sure, I don't have a strong opinion like you on that one

- I'll let that be something that you and the developer can decide and find the best way, maybe he is open to make them separate commands instead

**Comment by Aayush9029** [comment]
@pernielsentikaer I have changed the folder and your changes as requested.

I see you and @unnamedd seem to have a discussion going on if  this should be a script instead.

Eventhough I think this being on Store makes it easier to install, update. I'll leave the decision to you guys!

**Comment by unnamedd** [comment]
Hey @Aayush9029,

actually no, my point here it is just regarding the User Experience and in the end, the decision will be yours. In my opinion, the best UX here, would be exactly the one we have when we use Script Commands, which is a quick access to each command. Let me give you an example of my Raycast Preferences:
image

The `Open from Clipboard` is something that I use multiple times a day and makes sense for me to access it really after I took a screenshot (which in my case is copied directly to the clipboard), however, in your current implementation, this behaviour is not possible, because you have only one list with all the commands at the same place and a nice good UX, would be to have multiple commands.

What I mean?
Instead of having only the `index.ts` which list (`<List />`) all these options here:
image
would be better if you create multiple `.tsx` files like:
- capture-fullscreen.tsx
- capture-area.tsx
- open-from-clipboard.tsx

and so on..
In that way, you will have multiple commands separated (like on [Script Commands folder](https://github.com/raycast/script-commands/tree/master/commands/apps/cleanshot)) which will make possible to have a quick access and customized using an _Alias_ or _Hotkey_, like I presented in my first screenshot here.

Also, in that way, you won't need to have a view for each command, which you can set in your `packages.json` for each command `"mode": "no-view"`. 

Now you will may ask: _"Yes, Thiago, but then I won't have a list of all commands available in my extension."_

If you have that though, remember that Raycast will show all commands available, the user will just need to type "cleanshot" in the root search and your commands will be there as well.

To me, the biggest and most important thing is to make use of all the benefits provided by Raycast and having all the commands being listed inside of your unique view extension, won't allow customization or an easy way to quick use the most appropriated command the user needs.

**Comment by Aayush9029** [comment]
@unnamedd  I've moved each commands into separate .tsx files

AShot-2021 12 29-16 40 50

Screen Shot 2021-12-29 at 4 41 40 PM

I am wondering if there's a better way of sending urlsceheme commands instead of using macOSes  "open" command.

**Comment by es183923** [comment]
`CleanShotX` should be formatted as `CleanShot X`. Note the space between `CleanShot` and `X`.

**Comment by Aayush9029** [comment]
> CleanShotX

I thought so too, but it's named CleanShot X *with a space*

AShot-2021 12 29-18 58 34@2x

I'll  gladly remove a space if that's bothering your though!

**Comment by es183923** [comment]
No, you should add the space that you don't currently have.

**Comment by Aayush9029** [comment]
> No, you should add the space that you don't currently have.

Oop. In the title or folder name or somewhere else :)

**Comment by pernielsentikaer** [comment]
> Oop. In the title or folder name or somewhere else :)

In the subtitles of the commands

**Comment by unnamedd** [comment]
> @unnamedd I've moved each commands into separate .tsx files

Really nice @Aayush9029, now I guess it will make a huge success among the users, because now they can use each command separately. 

> I am wondering if there's a better way of sending urlsceheme commands instead of using macOSes "open" command.

As far as I saw, for example on Bear extension, the developer used a dependency called `open`, you won't have much where else to go. My suggestion, install that dependency and use it. Check [this](https://github.com/raycast/extensions/blob/main/extensions/bear/src/grab-url.tsx#L17) and [this](https://github.com/raycast/extensions/blob/main/extensions/bear/package.json#L37).

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:12]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:20]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:28]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:36]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:44]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:52]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:60]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:68]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:76]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:84]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:83]:
```suggestion
            "title": "Open From Clipboard",
```
Documentation: 
- [Extensions and Commands Naming](https://developers.raycast.com/basics/prepare-an-extension-for-store#extensions-and-commands-naming)
  - Extension and command titles should follow the Title Case convention

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:92]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:93]:
```suggestion
            "description": "Toggle Desktop Icons Visibility",
```

Documentation: 
- [Extensions and Commands Naming](https://developers.raycast.com/basics/prepare-an-extension-for-store#extensions-and-commands-naming)
  - Extension and command titles should follow the Title Case convention

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:100]:
```suggestion
            "subtitle": "CleanShot X",
```

File: extensions/cleanshotx/package.json
Comment (unnamedd) [inline extensions/cleanshotx/package.json:139]: As far as I could see, those information are not supported by Extension API, for that reason, I guess it will be irrelevant here. Also, in the end, the repository (for example) won't be the one pointed in the `repository` key, but for the repository where you are creating your PR (this one here).

**Review by unnamedd — CHANGES_REQUESTED** [review]

**Review by unnamedd — APPROVED** [review]
Hey @Aayush9029,
Your extension looks very nice to me now!

Thanks for your patience with all the requests and the extension itself!

**Comment by Aayush9029** [comment]
> Hey @Aayush9029, Your extension looks very nice to me now!
> 
> Thanks for your patience with all the requests and the extension itself!

No worries! Thanks for the amazing feedback!

**Review by pernielsentikaer — APPROVED** [review]
Hi @Aayush9029 

Congratulations with the extension, it turned out quite nice, and thanks to @unnamedd for the help turning it even better 💪 

I'm going to accept it and then will my colleague do a final review before it's finally merged 💎

**Comment by Aayush9029** [comment]
I've added icons and one more command (pin)  and that's about it though!

AShot-2021 12 31-13 33 33@2x

AShot-2021 12 31-13 35 30@2x

Looks super neat!

**Review by pernielsentikaer — APPROVED** [review]
Still accepted 😄

File: extensions/cleanshotx/src/annotate.tsx
Comment (mathieudutour) [inline extensions/cleanshotx/src/annotate.tsx:2]: have you run `rpm run lint`? I'm pretty sure prettier should rewrite this to `import { closeMainWindow } from ...`

File: extensions/cleanshotx/src/annotate.tsx
Comment (mathieudutour) [inline extensions/cleanshotx/src/annotate.tsx:8]: It should also remove that extra white line.

**Review by mathieudutour — COMMENTED** [review]
This looks really cool, well done!

Could you have a quick look at the 2 comments about code style? It's very minor and we should automate those check (I'll have a look tomorrow)

I'll merge the PR after that

File: extensions/cleanshotx/src/annotate.tsx
Comment (Aayush9029) [reply extensions/cleanshotx/src/annotate.tsx]: > have you run `rpm run lint`? I'm pretty sure prettier should rewrite this to `import { closeMainWindow } from ...`

Weird, linter didn't catch that, (perhaps my linter is not acting sanely)  here's my before and after!
AShot-2022 01 03-16 14 16@2x

Anyways, I've added and removed spaces as you requested!

**Review by mathieudutour — APPROVED** [review]
Awesome, thanks!

**Pull request merged**
